monitoring/uss_qualifier/suites/faa/uft/design/README.md

```diff
@@ -0,0 +1,87 @@
+    ```bash
+   ./run_locally.sh configurations.dev.faa.uft.local_message_signing
+   ```
+
+## Results
+SCD tests report is generated under [uss_qualifier](../../../../../uss_qualifier).
+The message signing results will be in the report created for the overall run - report.json - under the `report.messagesigning` section. This section will have an object titled `findings`, with two sections. The first section, `interactions`, is a list of all the interactions associated with each test. An example of an `interaction` is an incoming request from the USS under test, or an outgoing request to the USS under test. The second section, `issues`, will hold any message signing issues generated by the analysis of incoming requests and received responses. Some examples of issues that could be discovered are listed and detailed below. The current specification used for http message signing can be found [here](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-11).  
+
+1. Missing message signing headers - The first step of message signing analysis is to check for the existence of all of the required message signing headers. These headers are `content-digest`, `utm-message-signature`, `utm-message-signature-input`, and `x-utm-jws-header`. If any of these headers are missing in incoming requests to the mock_uss, or are missing in received responses from the USS under test, then that will generate an issue.
+2. Invalid content digests - Once all message signing headers are accounted for, the sha-512 hash on either the request/response body (depending on what is being analyzed), is taken, and this value is compared to the one received in the `content-digest` header. If these values differ, this will generate an issue showing the difference in the values. 
+3. Invalid signature - After the content-digest check is performed, the `utm-message-signature` is analyzed to ensure that it is a valid signature. To do this check, the signature base is created, and the public key served in the `x5u` field within the `x-utm-jws-header` is retreived. Using the value in the `utm-message-signature` header field, the public key, and the recreated signature base, it is determined whether or not the signature is valid.
+
+
+### Positive tests -
+A valid set of private/public keys are provided for message signing analysis under the [monitioring/monitorlib/messagesigning/keys] folder. The valid key pair, `mock_faa_priv.pem`/`mock_faa_pub.der`, is used by default. The public key is served under the mock_uss endpoint `/mock/scd/.well-known/uas-traffic-management/mock_pub.der`, and can be retreived by the USS under test in order for it to validate the mock_uss responses. This public key was provided by the FAA and will pass SCVP validation. To ensure that this valid keypair is used, make sure that the `USE_VALID_KEY_PAIR` in `run_locally_msgsigning.sh` is set to "true". 
```

Comment:
How about:

```suggestion
A set of private/public keys are provided for use by mock_uss in message signing analysis under the [monitioring/mock_uss/messagesigning/keys] folder. This key pair, `mock_faa_priv.pem`/`mock_faa_pub.der`, is used by mock_uss by default. The public key is served under the mock_uss endpoint `/mock/msgsigning/.well-known/uas-traffic-management/pub.der`, and can be retrieved by the USS under test in order for it to validate the mock_uss responses. This public key was provided by the FAA and will pass SCVP validation for the UFT activity.
```

I've removed "valid" as I don't think that's well-defined here.  I think "key pair" already implies that the public key matches the private key, and we say the public key will be recognized by SCVP in UFT; are there any other characteristics that make a key pair "valid"?

Since it seems that the only purpose of this sample key pair will be for mock_uss's use, it seems like they should be stored in the mock_uss folder.  Or, if they will be used for one or more other purposes, what will those purposes be?

I've changed `mock_pub.der` to simply `pub.der` because the usage intent of the content (to validate message signatures) shouldn't depend on the content itself (whether the certificate is a mock, a real one, an invalid one, etc).

Since the `.well-known/uas-traffic-management/pub.der` endpoint is specific to the `msgsigning` mock_uss capability, it should be hosted under that prefix rather than `scd`.

I've removed the last sentence because we previously say this key pair will be used by default, so there should be no need to have a separate setting to select it.  Moreover, since we will want to change key pairs during test to validate behavior with invalid key pairs (based on previous discussions), key pair selection should be accomplished via `msgsigning` endpoint rather than environment variable.

Reply:
` are there any other characteristics that make a key pair "valid"?`

The only thing I can think of outside of the public key passing `SCVP` validation would be the private key needs to be one that can be used to sign with the [RSASSA-PKCS1-v1_5 w/ SHA-256 algorithm](https://www.ietf.org/archive/id/draft-ietf-httpbis-message-signatures-11.html#name-rsassa-pkcs1-v1_5-using-sha)


`Since it seems that the only purpose of this sample key pair will be for mock_uss's use, it seems like they should be stored in the mock_uss folder. Or, if they will be used for one or more other purposes, what will those purposes be?`

The public key will only be used by `mock_uss` when it signs its responses, and the private key will be used to sign outgoing requests to the USS under test by the `AuthAdapter`. (There was some [discussion](https://github.com/interuss/dss/pull/897#discussion_r1041432715) on if this is the right place for this code, when the time comes to implement this actual feature we can revisit this discussion then, I would think. But because it's currently used in two different places I don't think the keys should be strictly under the `mock_uss` package. Showing its use in our current implementation [here](https://github.com/nasa/dss/blob/uss-qual-test-ext-msg-sign/monitoring/mock_uss/scdsc/routes_scdsc.py#L165) and [here](https://github.com/nasa/dss/blob/uss-qual-test-ext-msg-sign/monitoring/monitorlib/infrastructure.py#L61) for your reference.

Reply:
Forgot to include [this](https://github.com/nasa/dss/blob/uss-qual-test-ext-msg-sign/monitoring/messagesigning/message_signer.py#L41). The `signer` referenced is shown in more detail. Implementation will change based on these design modifications, but the private key will still be used in multiple places.

Reply:
Ok, so it sounds like the base `AuthAdapter` class will be modified to add additional headers to outgoing messages as part of the authorization being applied by the AuthAdapter -- I think that makes sense (though we'll want to change the AuthAdapter description to clarify its broader scope when those changes are made).  It would probably be helpful to write a summary of how that will be incorporated, but not necessary for this PR.  If I understand correctly, then these example auth specs:

* `NoAuth()`
* `ServiceAccount(https://example.com/token, ~/credentials/account.json)`
* `DummyOAuth(http://localhost:8085/token, sub=fake_uss)`

...would become something like these (when wanting to use UFT-style message signing):

* `NoAuth(message_signing=UFT(key_path=/auth/uss1.key, cert_url=https://uss1.example.com/uss1.der))`
* `ServiceAccount(https://example.com/token, ~/credentials/account.json, message_signing=UFT(key_path=/auth/uss1.key, cert_url=https://uss1.example.com/uss1.der))`
* `DummyOAuth(http://localhost:8085/token, sub=fake_uss, message_signing=UFT(key_path=/auth/uss1.key, cert_url=https://uss1.example.com/uss1.der))`

...and probably `SignedRequest` would then be modified to let the message signing functionality of the base AuthAdapter replace the custom logic currently in `SignedRequest` such that the auth spec

`SignedRequest(https://example.com/oauth/token, client_id=uss1.com,
   key_path=/auth/uss1.key, cert_url=https://uss1.com/uss1.der)`

...would likely become something like

* `FIMSRequest(https://example.com/oauth/token, message_signing=UFT(key_path=/auth/uss1.key, cert_url=https://uss1.com/uss1.der))`

If that understanding isn't correct, can we add something to the documentation describing the actual intention?  If it is correct, then it sounds like we will be using `mock_faa_priv.pem`/`mock_faa_pub.der` in roughly the same way as we use `build/test-certs/auth2.key`/`build/test-certs/auth2.pem` -- can we plan to put these two keys in test-certs then?  Revised suggestion:

> A set of private/public keys are provided for use by mock_uss in message signing analysis under the [build/test-certs/message-signing] folder. This key pair, `mock_faa_priv.pem`/`mock_faa_pub.der`, is used by mock_uss (when requested from an instance with the capability enabled) to both sign responses to incoming messages, and to sign requests of outgoing messages. When the message signing mock_uss capability is enabled, the public key is served under the mock_uss endpoint `/mock/msgsigning/.well-known/uas-traffic-management/pub.der`, and can be retrieved by the USS under test in order for it to validate the mock_uss responses, per UFT message signing requirements. This public key was provided by the FAA and will pass SCVP validation for the UFT activity.

Reply:
That is the intention, making sure to save this comment to come back to when we actually do the work. The only thing I'd disagree with is the naming of `SignedRequest`, I actually think we can keep that name. `FimsRequest` is a bit too narrow. It won't just be the FIMS server that would received signed requests, but also other USSs. `SignedRequest`, in my opinion, encapsulates this intention well.  Updated this paragraph with a minor edit, mock_uss signs it responses but it is not the module that signs outgoing requests, so I made sure to specify that `AuthAdapter` will be taking care of the signing of outgoing requests. Pushing shortly

Reply:
Nothing in this PR depends on `SignedRequest` vs `FIMSRequest` so I'll resolve the thread (feel free to unresolve to continue discussion, even after this PR's submission, or we can discuss further when we get to that later PR), but just FYI: my thought around the difference was that that particular AuthAdapter will no longer sign requests, right?  Message signatures will be done at the base `AuthAdapter` level?  In that case, it seems like there is no "signed request", but rather simply a request to FIMS (a request formatted the way FIMS wants it to be formatted), and the required signature will be attached at the AuthAdapter level, the same way message signatures could be attached (or not attached) to any of the other particular concrete AuthAdapter subclasses.